container/multi-value-slice/multi_value_slice.go

```diff
@@ -0,0 +1,380 @@
+		for _, v := range item.Values {
+			if v.val == val {
+				v.ids = append(v.ids, obj.Id())
+				newValue = false
+				break
+			}
+		}
+		if newValue {
+			item.Values = append(item.Values, &Value[V]{val: val, ids: []uint64{obj.Id()}})
+		}
+	}
+
+	return nil
+}
+
+// Append adds a new item to the input object.
+func (s *Slice[V, O]) Append(obj O, val V) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if len(s.appendedItems) == 0 {
+		s.appendedItems = append(s.appendedItems, &MultiValue[V]{Values: []*Value[V]{{val: val, ids: []uint64{obj.Id()}}}})
+		s.cachedLengths[obj.Id()] = len(s.sharedItems) + 1
+		return
+	}
+
+	for _, item := range s.appendedItems {
+		found := false
+		for _, v := range item.Values {
+			found = compareIds(v.ids, obj.Id(), nil)
+			if found {
+				break
+			}
+		}
+		if !found {
+			newValue := true
+			for _, v := range item.Values {
+				if v.val == val {
+					v.ids = append(v.ids, obj.Id())
+					newValue = false
+					break
+				}
+			}
+			if newValue {
+				item.Values = append(item.Values, &Value[V]{val: val, ids: []uint64{obj.Id()}})
+			}
+
+			l, ok := s.cachedLengths[obj.Id()]
+			if ok {
+				s.cachedLengths[obj.Id()] = l + 1
+			} else {
+				s.cachedLengths[obj.Id()] = len(s.sharedItems) + 1
+			}
+
+			return
+		}
+	}
+
+	s.appendedItems = append(s.appendedItems, &MultiValue[V]{Values: []*Value[V]{{val: val, ids: []uint64{obj.Id()}}}})
+
+	s.cachedLengths[obj.Id()] = s.cachedLengths[obj.Id()] + 1
+}
+
+// Detach removes the input object from the multi-value slice.
+// What this means in practice is that we remove all individual and appended values for that object and clear the cached length.
+func (s *Slice[V, O]) Detach(obj O) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	for i, ind := range s.individualItems {
+		for vi, v := range ind.Values {
+			found := compareIds(v.ids, obj.Id(), func(index int) {
+				if len(v.ids) == 1 {
+					if len(ind.Values) == 1 {
+						delete(s.individualItems, i)
+					} else {
+						ind.Values = append(ind.Values[:vi], ind.Values[vi+1:]...)
+					}
+				} else {
+					v.ids = append(v.ids[:index], v.ids[index+1:]...)
+				}
+			})
+			if found {
+				break
+			}
+		}
+	}
+
+	for _, item := range s.appendedItems {
+		found := false
+		for vi, v := range item.Values {
+			found = compareIds(v.ids, obj.Id(), func(index int) {
+				if len(v.ids) == 1 {
+					item.Values = append(item.Values[:vi], item.Values[vi+1:]...)
+				} else {
+					v.ids = append(v.ids[:index], v.ids[index+1:]...)
+				}
+			})
+			if found {
+				break
+			}
+		}
+		if !found {
+			// This is an optimization. If we didn't find an appended item at index i,
+			// then all larger indices don't have an appended item for the object either.
+			break
+		}
+	}
+
+	delete(s.cachedLengths, obj.Id())
+}
+
+// compareIds executes a custom action when the wanted ID is found. It returns whether the ID was found.
+func compareIds(ids []interfaces.Id, wanted interfaces.Id, action func(index int)) bool {
```

Comment:
The side-effect callback is a bit hard to follow when using this. Why not just compare ids and return whether it was found? and then do the action afterwards from the caller

Reply:
Maybe should be called `containsId`

Reply:
I removed the callback. I now return the found index from the function and use that index for subsequent actions.